File: daemon/protocol.py
"""Shared message schemas for the approval protocol.

Two independent wire protocols use these builders:
- The local-network TCP link between approve_daemon.py and the phone
  (hello/hello_ack/request/response).
- The local AF_UNIX link between hooks/pretooluse_approve.py (the hook) and
  approve_daemon.py.

Framing for both is one JSON object per line: json.dumps(msg) + "\\n". Never
pass indent= to json.dumps here - it must stay single-line, since embedded
newlines in string values are escaped by json.dumps and that's what makes
newline-delimited framing safe.

The Android app's model/Protocol.kt mirrors the TCP link messages by hand;
keep the two in sync when changing field names here.
"""
import json
import os

# v2: pairing payload switched from Bluetooth mac/channel to host/port
# (RFCOMM replaced with local-network TCP so the daemon works on macOS too,
# not just Linux/BlueZ). The field rename alone already makes an old QR
# code decode-fail cleanly against the new app (and vice versa); this bump
# just keeps the marker meaningful for anyone diffing payloads by hand.
#
# v3: the link is now encrypted (see daemon/secure_channel.py) - every
# connection does an ECDH key exchange before the hello/hello_ack/request/
# response messages below, which now travel AES-GCM-encrypted instead of
# in the clear. An old plaintext-only app/daemon talking to a new encrypted
# one fails the handshake outright (it never sends/expects the initial kex
# message), rather than silently working unencrypted.
#
# v4: Bluetooth RFCOMM is back as a second transport, running alongside TCP
# rather than replacing it (see daemon/phone_link.py) - the pairing payload
# now optionally carries bt_mac/bt_channel so the phone can race a Bluetooth
# connection attempt against the TCP one and use whichever completes its
# handshake first. Both new fields are optional/additive (unlike the v1->v2
# rename), so a v3 app or daemon still talks TCP-only with a v4 peer.
#
# v5: classic Bluetooth RFCOMM *server* mode isn't reliably available to
# third-party apps on macOS (bluetoothd owns the RFCOMM mux and doesn't
# expose server-mode registration the way BlueZ does on Linux) - see
# daemon/bt_backend_macos.py. macOS now runs a BLE peripheral/GATT server
# instead and signals this via the optional bt_le flag; Linux is unaffected
# and keeps emitting bt_mac/bt_channel exactly as in v4. Also additive, so a
# v4 peer just doesn't see the new field and stays TCP-only against a macOS
# v5 daemon (which is a no-op change from today, since macOS Bluetooth
# effectively never worked before this).
#
# v6: adds the "cancel" message (daemon -> phone), sent when the daemon
# gives up waiting on a request past REQUEST_TIMEOUT_SECONDS - see
# build_cancel() below. Purely additive (a new message type an old app
# never sees), so a pre-v6 app just keeps showing a request card the
# daemon has already abandoned until the phone's own request notification/
# card is dismissed by hand; the app-side self-expiry timer that mirrors
# REQUEST_TIMEOUT_SECONDS (DaemonLinkManager.kt) is what actually needs the
# new app build, not this wire change itself.
PROTOCOL_VERSION = 6
TOOL_INPUT_TRUNCATE = 1200
STATE_DIR_NAME = ".phone-ai-approve"
RUNTIME_DIR_NAME = "phone-ai-approve"

# How long the daemon waits for the phone to answer a request before giving
# up and telling the hook to fail open - see approve_daemon.py's
# handle_local_request(). Single source of truth for both that wait and the
# "cancel" message sent to the phone when it expires, so the two can never
# drift apart; DaemonLinkManager.kt's REQUEST_TIMEOUT_SECONDS mirrors this
# value by hand (see its own comment) since the phone has no way to read
# this file.
REQUEST_TIMEOUT_SECONDS = 100.0

# Fixed RFCOMM channel the daemon's Bluetooth listener binds to and the app
# connects to directly (by channel number, not SDP lookup) - see
# daemon/phone_link.py's LinuxBtPhoneLink and daemon/bt_backend_macos.py for
# why a fixed channel sidesteps needing a full SDP server on Linux. Kept here
# as the single source of truth so payload generation (pairing.py) and the
# actual bind/publish (phone_link.py / bt_backend_macos.py) never drift apart.
BT_CHANNEL = 4

# BLE GATT UUIDs for the macOS Bluetooth transport (see
# daemon/bt_backend_macos.py's CBPeripheralManager-based server and
# BluetoothLeClient.kt's matching central). Fixed constants, same spirit as
# BT_CHANNEL above - both sides know them ahead of time, no discovery/
# negotiation needed. RX is phone->daemon (Write, with response - Write-
# without-response requests never reach peripheralManager:didReceiveWriteRequests:,
# so it's deliberately not offered); TX is daemon->phone (Notify). A peer
# that happens to advertise/scan for the same service UUID without knowing
# our pairing token just fails the handshake and gets retried, exactly like
# a stray connection to TCP_PORT already does today - this is not a new
# class of exposure.
BT_LE_SERVICE_UUID = "7f3a2b1a-0001-4c7a-9c1f-6f3f2b6a8e11"
BT_LE_RX_CHAR_UUID = "7f3a2b1a-0002-4c7a-9c1f-6f3f2b6a8e11"
BT_LE_TX_CHAR_UUID = "7f3a2b1a-0003-4c7a-9c1f-6f3f2b6a8e11"

# Actions a phone response (or a local daemon->hook response) may carry.
PHONE_ACTIONS = {"allow", "allow_always", "deny", "other"}
# Local daemon->hook responses additionally cover daemon-side failure modes;
# both mean the hook must print nothing and fail open.
LOCAL_ACTIONS = PHONE_ACTIONS | {"timeout", "no_phone"}


def state_dir() -> str:
    d = os.path.join(os.path.expanduser("~"), STATE_DIR_NAME)
    os.makedirs(d, exist_ok=True)
    return d


def runtime_dir() -> str:
    # Codex can run hooks in a sandbox where ~/.phone-ai-approve is readable but
    # AF_UNIX connect() to a socket there is denied. /tmp is intentionally
    # writable/visible to those hooks, so keep only the local relay socket here.
    d = os.path.join("/tmp", f"{RUNTIME_DIR_NAME}-{os.getuid()}")
    os.makedirs(d, mode=0o700, exist_ok=True)
    try:
        os.chmod(d, 0o700)
    except OSError:
        pass
    return d


def daemon_sock_path() -> str:
    return os.path.join(runtime_dir(), "daemon.sock")


def file_relay_dir() -> str:
    d = os.path.join(runtime_dir(), "requests")
    os.makedirs(d, mode=0o700, exist_ok=True)
    try:
        os.chmod(d, 0o700)
    except OSError:
        pass
    return d


def encode(msg: dict) -> bytes:
    return (json.dumps(msg) + "\n").encode("utf-8")


def summarize_tool_input(tool_input) -> str:
    if isinstance(tool_input, dict):
        text = "\n".join(f"{k}: {v}" for k, v in tool_input.items())
    else:
        text = str(tool_input)
    return text[:TOOL_INPUT_TRUNCATE]


# ---- TCP link messages (daemon <-> phone) ----

def build_hello(token: str) -> dict:
    return {"type": "hello", "tok": token}


def build_hello_ack(ok: bool) -> dict:
    return {"type": "hello_ack", "ok": ok}


def build_request(req_id, session_id, tool_name, tool_input, cwd, ts, options=None) -> dict:
    msg = {
        "type": "request",
        "req_id": req_id,
        "session_id": session_id,
        "tool_name": tool_name,
        "tool_input": summarize_tool_input(tool_input),
        "cwd": cwd,
        "ts": ts,
    }
    # Additive, like bt_mac/bt_channel on the pairing payload: a proposed-answer
    # tool (Claude Code's AskUserQuestion; see hooks/pretooluse_approve.py) sets
    # this so the phone can offer the actual choices as buttons instead of a
    # generic Allow/Allow always/Deny that doesn't correspond to any real answer.
    # Omitted (not just null) for every ordinary tool call, so an old app build
    # that doesn't know this field still renders those requests exactly as before.
    if options:
        msg["options"] = options
    return msg


def build_phone_response(req_id, action, reply=None) -> dict:
    if action not in PHONE_ACTIONS:
        raise ValueError(f"invalid phone action: {action!r}")
    msg = {"type": "response", "req_id": req_id, "action": action}
    if reply is not None:
        msg["reply"] = reply
    return msg


def build_notify(session_id, cwd, message, ts) -> dict:
    return {
        "type": "notify",
        "session_id": session_id,
        "cwd": cwd,
        "message": message,
        "ts": ts,
    }


def build_cancel(req_id) -> dict:
    """Daemon -> phone: this request is dead (timed out, or the hook's
    local connection dropped before an answer came back), so the phone
    should drop its card/notification for req_id immediately rather than
    leaving it stuck until its own self-expiry timer catches up."""
    return {"type": "cancel", "req_id": req_id}


# ---- Local AF_UNIX messages (hook <-> daemon) ----

def build_local_request(req_id, session_id, tool_name, tool_input, cwd, options=None) -> dict:
    msg = {
        "type": "request",
        "req_id": req_id,
        "session_id": session_id,
        "tool_name": tool_name,
        "tool_input": tool_input,
        "cwd": cwd,
    }
    if options:
        msg["options"] = options
    return msg


def build_local_notify(session_id, cwd, message) -> dict:
    return {
        "type": "notify",
        "session_id": session_id,
        "cwd": cwd,
        "message": message,
    }


def build_local_response(action, reason="") -> dict:
    if action not in LOCAL_ACTIONS:
        raise ValueError(f"invalid local action: {action!r}")
    return {"action": action, "reason": reason}

File: app/src/main/java/com/phoneapprove/app/model/Protocol.kt
package com.phoneapprove.app.model

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.decodeFromJsonElement
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

/** Mirrors daemon/protocol.py's message schemas - keep the two in sync. */

@Serializable
data class QrPairingPayload(
    val v: Int,
    val id: String,
    val host: String,
    val port: Int,
    val tok: String,
    val name: String,
    // Optional: absent (null) when the daemon had no usable Bluetooth adapter
    // at pairing time, in which case this pairing is TCP-only. Both default
    // to null so a v3 (TCP-only) payload still parses fine here. Classic-
    // RFCOMM-only (a Linux daemon); a macOS daemon sets bt_le instead (see
    // below), never these two.
    val bt_mac: String? = null,
    val bt_channel: Int? = null,
    // True when the daemon is macOS and runs a BLE peripheral/GATT server
    // (see daemon/bt_backend_macos.py) rather than classic RFCOMM - classic
    // RFCOMM server mode isn't reliably available to third-party apps on
    // macOS. Defaults to false so a pre-v5 payload still parses fine here.
    val bt_le: Boolean = false,
)

/** Mirrors daemon/protocol.py's BT_LE_* constants - fixed GATT UUIDs for the
 * macOS BLE transport, see BluetoothLeClient.kt. */
object BleUuids {
    const val SERVICE = "7f3a2b1a-0001-4c7a-9c1f-6f3f2b6a8e11"
    const val RX_CHARACTERISTIC = "7f3a2b1a-0002-4c7a-9c1f-6f3f2b6a8e11" // phone -> daemon (write)
    const val TX_CHARACTERISTIC = "7f3a2b1a-0003-4c7a-9c1f-6f3f2b6a8e11" // daemon -> phone (notify)
}

@Serializable
data class HelloMessage(val type: String = "hello", val tok: String)

@Serializable
data class HelloAckMessage(val type: String = "hello_ack", val ok: Boolean)

@Serializable
data class RequestMessage(
    val type: String = "request",
    val req_id: String,
    val session_id: String,
    val tool_name: String,
    val tool_input: String,
    val cwd: String,
    val ts: Double,
    // Present only for a proposed-answer tool (Claude Code's AskUserQuestion -
    // see hooks/pretooluse_approve.py's _question()); null for every ordinary
    // tool call, which keeps showing the plain Allow/Allow always/Deny row.
    val options: List<String>? = null,
)

@Serializable
data class ResponseMessage(
    val type: String = "response",
    val req_id: String,
    val action: String,
    val reply: String? = null,
)

@Serializable
data class NotifyMessage(
    val type: String = "notify",
    val session_id: String,
    val cwd: String,
    val message: String,
    val ts: Double,
)

@Serializable
data class CancelMessage(
    val type: String = "cancel",
    val req_id: String,
)

// encodeDefaults=true is required: HelloMessage/ResponseMessage's `type`
// field has a default value ("hello"/"response"), and kotlinx.serialization
// omits fields that equal their default unless told otherwise - confirmed
// via hands-on testing, where the daemon received {"tok":"..."} with no
// "type" key at all and silently rejected it as a handshake mismatch.
val protocolJson = Json { ignoreUnknownKeys = true; encodeDefaults = true }

sealed class IncomingMessage {
    data class Ack(val ok: Boolean) : IncomingMessage()
    data class Req(val message: RequestMessage) : IncomingMessage()
    data class Notify(val message: NotifyMessage) : IncomingMessage()
    data class Cancel(val reqId: String) : IncomingMessage()
    object Unknown : IncomingMessage()
}

fun parseIncoming(line: String): IncomingMessage {
    return try {
        val obj = protocolJson.parseToJsonElement(line).jsonObject
        when (obj["type"]?.jsonPrimitive?.contentOrNull) {
            "hello_ack" -> IncomingMessage.Ack(protocolJson.decodeFromJsonElement(HelloAckMessage.serializer(), obj).ok)
            "request" -> IncomingMessage.Req(protocolJson.decodeFromJsonElement(RequestMessage.serializer(), obj))
            "notify" -> IncomingMessage.Notify(protocolJson.decodeFromJsonElement(NotifyMessage.serializer(), obj))
            "cancel" -> IncomingMessage.Cancel(protocolJson.decodeFromJsonElement(CancelMessage.serializer(), obj).req_id)
            else -> IncomingMessage.Unknown
        }
    } catch (e: Exception) {
        IncomingMessage.Unknown
    }
}

fun parseQrPayload(text: String): QrPairingPayload? {
    return try {
        protocolJson.decodeFromString(QrPairingPayload.serializer(), text)
    } catch (e: Exception) {
        null
    }
}
